matsuda-reversi/reversi4.html

```diff
@@ -0,0 +1,181 @@
+          ctx.moveTo(0, BLOCK_SIZE * i);
+          ctx.lineTo(FIELD_SIZE, BLOCK_SIZE * i);
+          ctx.moveTo(BLOCK_SIZE * i, 0);
+          ctx.lineTo(BLOCK_SIZE * i, FIELD_SIZE);
+          ctx.closePath();
+          ctx.stroke();
+        }
+      };
+      // banmenデータから石を描画
+      banmen.refresh = function () {
+        for (let x = 0; x < 8; x++) {
+          for (let y = 0; y < 8; y++) {
+            drawStone(x, y, this[y][x]);
+          }
+        }
+      };
+      // 特定の方向に配置してあるひっくり返せる石の情報を集める
+      banmen.collectStones = function (x, y, color, direction) {
+        const reverseStoneColor = color === WHITE ? BLACK : WHITE; // 手番ではない石の色
+        const stones = []
+        // 最大7回繰り返し、石の情報を集める
+        for (let i = 0; i < 7; i++) {
+          x += direction.x;
+          y += direction.y;
+
+          // 盤面の外なので、ループ処理を抜ける
+          if (x > 7 || x < 0 || y > 7 || y < 0) break;
+          // 何も置いてないマスがある場合、ループ処理を抜ける
+          if (this[y][x] === 0) break;
+          // 石の情報を配列に追加
+          stones.push(new Stone(x, y, this[y][x]));
+          // 先頭以外で、自分の色が出たらその時点でループ処理を抜ける
+          if (this[y][x] === color) break;
+        }
+        
+        // 一つ以下しか配列がない場合、ひっくり返せない
+        // ひっくり返せない状況の場合は、空の配列を返す
+        if (stones.length <= 1) return [];
+        // 配列の先頭が相手の石の色でない場合、ひっくり返せない
+        if (stones[0].color !== reverseStoneColor) return []; 
+        // 末尾の石の色が自分と同じ色でない場合、ひっくり返せない
+        if (stones[stones.length - 1].color !== color) return [];
+
+        return stones;
+      }
+
+      banmen.drawReversiFiledLines();
+      banmen.refresh();
+      // 盤面データ ----------
+
+      class Stone {
+        constructor(x, y, color) {
+          this.x = x;
+          this.y = y;
+          this.color = color;
+        }
+      }
+
+      // 石がその位置に置けるかを調べる関数
+      function canPutStone(originX, originY, color, simulation = false) {
+        const canReverse = false; // ひっくり返せるかのフラグ
+        let allStones = []; // ひっくり返せる石を格納しておく配列
+
+        if(banmen[originY][originX] !== 0) return false; // 空じゃないマスには置けない
+        
+        // 石を置きたい場所の八方向それぞれについて石がどのように配置されているか調べる
+        directions.forEach((direction) => {
+          // 特定方向の石を配列に貯める
+          const stones = banmen.collectStones(originX, originY, color, direction);
+          // ひっくり返せる全ての石の配列を作る
+          //allStones = [...allStones, ...stones]
+          allStones = allStones.concat(stones)
+        });
+
+        // ひっくり返せる石の配列を見ての処理
+        if (simulation) {
+          // シミュレーション
+          // この機能を使って、ひっくり返せる位置をユーザーに知らせる仕組みが作れる
+          return allStones.length !== 0;
+        } else {
+          // 本番(実際に石を引っくり返す)
+          // ひっくり返せる石がない
+          if (allStones.length === 0) return false;
+
+          allStones.push(new Stone(originX, originY, color));
+          allStones.forEach(stone => drawStone(stone.x, stone.y, color)); // 自分の色にひっくりかえす
+          
+          // 手番の交代
+          changePlayer();
+          banmen.refresh();
+
+          return true;
```

Comment:
好みあるでしょうが、返せる石があるかどうかのbooleanを返すので、こうはどうでしょう
```suggestion
          const canReverse = allStones.length === 0
          if (canReverse) {
            allStones.push(new Stone(originX, originY, color));
            allStones.forEach(stone => drawStone(stone.x, stone.y, color)); // 自分の色にひっくりかえす
          
            // 手番の交代
            changePlayer();
            banmen.refresh();          
          }

          return canReverse;
```

Reply:
` const canReverse = allStones.length` こうだな